ethers-solc/src/artifacts/mod.rs

```diff
@@ -1828,6 +1829,81 @@ impl SourceFiles {
     }
 }
 
+pub trait Linkable {
+    /// Tries to link the bytecode object with the `file` and `library` name.
+    /// Replaces all library placeholders with the given address.
+    ///
+    /// Returns true if the bytecode object is fully linked, false otherwise
+    /// This is a noop if the bytecode object is already fully linked.
+    fn link(&mut self, file: impl AsRef<str>, library: impl AsRef<str>, address: Address) -> bool;
+    /// Links the bytecode object with all provided `(file, lib, addr)`
+    fn link_all<I, S, T>(&mut self, libs: I) -> bool
+    where
+        I: IntoIterator<Item = (S, T, Address)>,
+        S: AsRef<str>,
+        T: AsRef<str>,
+    {
+        for (file, lib, addr) in libs.into_iter() {
+            if self.link(file, lib, addr) {
+                return true
+            }
+        }
+        false
+    }
+
+    /// Same as `Bytecode::link` but with fully qualified name (`file.sol:Math`)
+    fn link_fully_qualified(&mut self, name: impl AsRef<str>, addr: Address) -> bool {
+        if let Some((file, lib)) = name.as_ref().split_once(':') {
+            self.link(file, lib, addr)
+        } else {
+            false
+        }
+    }
+    /// Links the bytecode object with all provided `(fully_qualified, addr)`
+    fn link_all_fully_qualified<I, S>(&mut self, libs: I) -> bool
+    where
+        I: IntoIterator<Item = (S, Address)>,
+        S: AsRef<str>,
+    {
+        for (name, addr) in libs.into_iter() {
+            if self.link_fully_qualified(name, addr) {
+                return true
+            }
+        }
+        false
+    }
+
+    /// Whether the object is fully linked or not.
+    fn is_linked(&self) -> bool {
+        !self.is_unlinked()
+    }
+    /// Whether the object is fully linked or not.
+    fn is_unlinked(&self) -> bool;
+}
+
+/// A map of library bytecode to the address it was deployed to
+pub type Dependencies = Vec<(Address, Bytecode)>;
+
+/// A map of artifact IDs to their [Dependencies].
+pub type LinkerOutput = BTreeMap<ArtifactId, Dependencies>;
+
+/// A linker function that is provided an [Artifact], a library's [Artifact],
+/// the `index` of the library denoting the deployment order
+/// for the library in respect to the provided contract.
+///
+/// The function should return the address at which the library is deployed.
+pub trait LinkerFn:
+    FnMut((&ArtifactId, &dyn Artifact), (&ArtifactId, &dyn Artifact), usize) -> Option<Address>
+{
```

Comment:
this looks pretty wild

not sure I understand what that's supposed to do

can't we do this

`trait Linker { fn link(....)}`

?

Reply:
it's needed (or something like it) for e.g. `forge test` where we basically just want to get the artifact that should be linked, what library we are linking with, and what order it has. using this we can compute the address of the library before it is deployed

if we don't have something like this we have to keep pretty much all linker logic we have in foundry rn